### PR TITLE
Added `state upgrade`

### PR DESCRIPTION
**Pull request by Naatan**
<table>
<td>
  <a href="https://activestatef.atlassian.net/browse/DX-2623" title="DX-2623" target="_blank">StoryDX-2623</a>  As a user I can run a command and see what updates are available for my project
  </td></table>
  <br />
 

Couple of notable supplemental changes to facilitate `state upgrade`:

- Refactored buildplan changesets (ie. `buildplan.Diff()`) to facilitate iteration of all changes regardless of type.
- Expansion of `--ts` values moved to runbit.
- `buildplan.*.Dependencies()` now take an extra argument that allows you to skip recursion for certain artifacts.
- Added `sliceutils.EqualValues` which lets you compare two slices regardless of their ordering.
- Added `output.Structured()` which allows you to print structured data without needing to create your own marshaller.
- Centralized the definition of tree symbols.

### CODE REVIEW

File: pkg/buildplan/artifact_test.go [hunk 1]
```diff
@@ -54,14 +54,13 @@ func TestArtifact_Dependencies(t *testing.T) {
 			want: []string{
 				"00000000-0000-0000-0000-000000000002",
 				"00000000-0000-0000-0000-000000000003",
-				"00000000-0000-0000-0000-000000000001",
```
Comment (Naatan) [inline pkg/buildplan/artifact_test.go:57]: The old assertion failed due to my fix here:

https://github.com/ActiveState/cli/pull/3467/files#diff-9a4a1ec82a2498e8c04daf3594acd7e9f1de41731d7ef7539864881784a5c5c9R270

Before that you would get artifacts that had technically already been "seen". ie. asking for deps on artifact ID `1` would return `1`. Which is not expected (and ironically; also a cycle).

File: internal/sliceutils/sliceutils.go [hunk 2]
```diff
@@ -112,3 +114,23 @@ func ToLookupMapByKey[T any, K string | int | strfmt.UUID](data []T, keyCb func(
 	}
 	return result
 }
+
+// EqualValues checks if two slices have equal values, regardless of ordering. This does not recurse into nested slices or structs.
+func EqualValues[S ~[]E, E cmp.Ordered](a, b S) bool {
```
Comment (MDrakos) [inline internal/sliceutils/sliceutils.go:119]: nit: Unit test for this function would be nice.

**Review by MDrakos — CHANGES_REQUESTED** [review]
This looks great to me, just some minor comments and one question.

I would also recommend running the `cve` integration tests as well as any that depend on change summaries just to ensure that the changes there didn't affect anything.

File: internal/locale/locales/en-us.yaml [hunk 3]
```diff
@@ -1549,3 +1549,35 @@ warn_additional_requirements:
...
+  other: No upgrades were found for your project. Note that requirements with fixed version numbers will never be upgraded.
+upgrade_aborted:
+  other: Upgrade aborted
+upgrade_success:
+  other: Upgrade completed
+upgrade_field_same:
+  other: "[CYAN]{{.V0}}[/RESET]"
+upgrade_field_change:
+  other: "[CYAN]{{.V0}}[/RESET] > [BOLD][ACTIONABLE]{{.V1}}[/RESET]"
+name:
```
[the remark this answers is not in the document]
Comment (Naatan) [reply internal/locale/locales/en-us.yaml:1566]: No, the resets act on everything. It's basically telling the shell "go back to normal rendering".

File: internal/runners/upgrade/upgrade.go [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+		return errs.Wrap(err, "Failed to clone build script")
+	}
+	ts, err := commits_runbit.ExpandTime(&params.Timestamp, u.prime.Auth())
+	if err != nil {
+		return errs.Wrap(err, "Failed to fetch latest timestamp")
+	}
+	bumpedBS.SetAtTime(ts)
+
+	if bumpedBS.AtTime().String() == localCommit.BuildScript().AtTime().String() {
+		panic(fmt.Sprintf("bumped buildscript is same as local commit, old: %s, new: %s", bumpedBS.AtTime(), localCommit.BuildScript().AtTime()))
```
[the remark this answers is not in the document]
Comment (Naatan) [reply internal/runners/upgrade/upgrade.go:116]: Woops, it is indeed leftover. Thanks for spotting!

**Review by MDrakos — APPROVED** [review]

**Pull request merged**